### PR TITLE
COMMERCE-10054 - OM - User without any permissions can add items to the cart

### PR DESCRIPTION
**Pull request by brianikim**
https://issues.liferay.com/browse/COMMERCE-10054

@andrea-ale-sbarra ,

Solution discussed with @drewbrokke.

@drewbrokke , let me know what you think about the new implementation. We only need to check whether the account is a businessAccount because:

1. businessAccount is the only account allowed in a B2B environment, therefore we do permission checks on those.
2. B2C only allows guest "account" or personalAccount. They can add to cart without any special permission checks.

To clarify, "adding to cart" creates an order, which is why the implementations are only for when an order is initially added.

Sincerely,
Brian

### CODE REVIEW

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:sf

**Comment by drewbrokke** [comment]
@brianikim I remember discussing that we should check the "guest checkout" preference on the given channel. I don't see that in this pull, is it no longer necessary?

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by drewbrokke** [comment]
@brianikim are there any integration tests to cover this scenario? If so, we should probably update them to reflect this change.

**Comment by brianikim** [comment]
@drewbrokke , no problem. I will take care of it.

**Comment by brianikim** [comment]
ci:test:relevant

**Review by drewbrokke — CHANGES_REQUESTED** [review]
@brianikim looking good, but I still have a couple of outstanding questions:

1. Shouldn't we provide test coverage to assert that a guest user _cannot_ create an order with a B2B account?
2. Is it out of scope of this ticket to consider whether or not the channel has guest checkout enabled? I remember that was part of our conversation, but I don't see it covered here.

Thanks!

**Comment by brianikim** [comment]
@drewbrokke , 

1. I will take care of the test coverage for Guest + B2B account.
2. It is out of scope. It may be added later, but I need to have a separate conversation for guest checkout scenario with the team about this.

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderServiceImpl.java [hunk 1]
```diff
@@ -86,6 +93,16 @@ public CommerceOrder addOrUpdateCommerceOrder(
...
+				commerceAccountId);
+
+			if (commerceAccount.isBusinessAccount()) {
+				_portletResourcePermission.check(
+					getPermissionChecker(),
+					commerceAccount.getCommerceAccountGroupId(),
+					CommerceOrderActionKeys.ADD_COMMERCE_ORDER);
+			}
+
 			_portletResourcePermission.check(
```
Comment (luca-pellizzon) [inline modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderServiceImpl.java:106]: This is executed even if the account is a business type. It is not clear to me why we need to check 2 different permissions.

**Review by luca-pellizzon — COMMENTED** [review]
@brianikim check my comment.
Also, was the user able to checkout the order or just add it to the cart?
If he can checkout the order, that's a worse issue we need to solve.

**Comment by brianikim** [comment]
@luca-pellizzon , updated PR with change requested.
@drewbrokke , new update:
1. Guest users can never be assigned B2B accounts, so the test is not necessary I think.
2. We are going to handle guestCheckout in a separate COMMERCE ticket most likely.

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by drewbrokke** [comment]
@brianikim don't we still need to cover the case where the user is a member of a B2B account, but doesn't have permission to add an order?

**Comment by brianikim** [comment]
@drewbrokke , thank you for clarifying. Your question prompted me to take a look at the test implementations again. I realized I implemented the withBuyerRole test incorrectly and adjusted accordingly wile also creating a new testcase for withoutBuyerRole. The new methods should satisfy your requests. Please take a look again.

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by luca-pellizzon** [comment]
@brianikim we need to give the "sales agent" role the ADD_COMMERCE_ORDER permission.
Please update the upgrade process to do so and also the "sales agent" role import.
After that please also run the test:commerce suite.
Many thanks.

**Comment by brianikim** [comment]
@luca-pellizzon , 

Just one question. I want to confirm whether we do any import logic on COMMERCE side for roles. The role imports should strictly be USM implementation right?

**Comment by brianikim** [comment]
@luca-pellizzon , I don't see a specific class dedicated to importing roles. I do see createRoles in [CPFileImporterImpl](https://github.com/liferay-commerce/liferay-portal/blob/1e38b8e6b2c0552018771f42878fb174ff93ec73/modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/internal/importer/CPFileImporterImpl.java#L157). If this is not the correct place, could you let me know the class and/or method that is handling the roles import?

**Comment by luca-pellizzon** [comment]
@brianikim check 
modules/apps/commerce/commerce-theme-minium/commerce-theme-minium-site-initializer/src/main/resources/com/liferay/commerce/theme/minium/site/initializer/internal/dependencies/roles.json

**Comment by brianikim** [comment]
@luca-pellizzon ,

I was aware of the json files, but I did not understand how it was being used. I looked into it further and I see how they are used.

I will update the PR accordingly. Thank you for the clarifications.

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:commerce

**Comment by brianikim** [comment]
Hello @luca-pellizzon , I introduced the new changes as you requested.

**Comment by luca-pellizzon** [comment]
@brianikim can you please update this one? many thanks

**Comment by luca-pellizzon** [comment]
ci:test:sf

**Comment by luca-pellizzon** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
@luca-pellizzon , could you review https://github.com/liferay-commerce/liferay-portal/pull/3083/commits/87fd5f333c4ce20643996bdf72fd31b6aae6d547 . It's a new commit introduced to disable addToCart button when the account is a business account and the user does not have ADD_COMMERCE_ORDER permission (Was requested by David Kelly).

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:commerce

**Review by luca-pellizzon — APPROVED** [review]

**Comment by luca-pellizzon** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:commerce

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
@luca-pellizzon , [9be264b](https://github.com/liferay-commerce/liferay-portal/pull/3083/commits/9be264b640fa8a4a541b33319ff8d77645e54921) needs to be reviewed. This is to address https://issues.liferay.com/browse/COMMERCE-10659 and https://issues.liferay.com/browse/COMMERCE-10660

**Comment by brianikim** [comment]
@victoroscuro , the testCases that involve Minium and adding-to-cart which are failing now all need to have proper permissions also as part of the testCases for them to pass.

**Review by luca-pellizzon — APPROVED** [review]

**Comment by brianikim** [comment]
https://github.com/liferay-commerce/liferay-portal/pull/3342

**Pull request closed without merging**